man/wat2wasm.1
.Dd $Mdocdate$
.Dt WABT 1
.Os
.Sh NAME
.Nm wat2wasm
.Nd translate from WebAssembly text format to the WebAssembly binary format
.Sh SYNOPSIS
.Nm wat2wasm
.Op options
.Ar filename
.Sh DESCRIPTION
.Nm
read a file in the wasm text format, check it for errors, and  convert it to the wasm binary format.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl Fl help
Print this help message
.It Fl Fl version
Print version information
.It Fl v , Fl Fl verbose
Use multiple times for more info
.It Fl d , Fl Fl dump-module
Print a hexdump of the module to stdout
.It Fl Fl disable-exceptions
Disable Experimental exception handling
.It Fl Fl disable-mutable-globals
Disable Import/export mutable globals
.It Fl Fl disable-saturating-float-to-int
Disable Saturating float-to-int operators
.It Fl Fl disable-sign-extension
Disable Sign-extension operators
.It Fl Fl disable-simd
Disable SIMD support
.It Fl Fl enable-threads
Enable Threading support
.It Fl Fl enable-function-references
Enable Typed function references
.It Fl Fl disable-multi-value
Disable Multi-value
.It Fl Fl disable-tail-call
Disable Tail-call support
.It Fl Fl disable-bulk-memory
Disable Bulk-memory operations
.It Fl Fl disable-reference-types
Disable Reference types (externref)
.It Fl Fl disable-annotations
Disable Custom annotation syntax
.It Fl Fl enable-code-metadata
Enable Code metadata
.It Fl Fl enable-gc
Enable Garbage collection
.It Fl Fl disable-memory64
Disable 64-bit memory
.It Fl Fl disable-multi-memory
Disable Multi-memory
.It Fl Fl disable-extended-const
Disable Extended constant expressions
.It Fl Fl disable-relaxed-simd
Disable Relaxed SIMD
.It Fl Fl enable-custom-page-sizes
Enable Custom page sizes
.It Fl Fl enable-compact-imports
Enable Compact import section
.It Fl Fl enable-wide-arithmetic
Enable Wide arithmetic
.It Fl Fl enable-all
Enable all features
.It Fl o , Fl Fl output=FILE
Output wasm binary file. Use "-" to write to stdout.
.It Fl r , Fl Fl relocatable
Create a relocatable wasm binary (suitable for linking with e.g. lld)
.It Fl Fl no-canonicalize-leb128s
Write all LEB128 sizes as 5-bytes instead of their minimal size
.It Fl Fl debug-names
Write debug names to the generated binary file
.El
.Sh EXAMPLES
parse test.wat and write to .wasm binary file with the same name
.Pp
.Dl $ wat2wasm test.wat
parse test.wat and write to binary file test.wasm
.Pp
.Dl $ wat2wasm test.wat -o test.wasm
parse spec-test.wast, and write verbose output to stdout (including
the meaning of every byte)
.Pp
.Dl $ wat2wasm spec-test.wast -v
.Sh SEE ALSO
.Xr wasm-interp 1 ,
.Xr wasm-objdump 1 ,
.Xr wasm-stats 1 ,
.Xr wasm-strip 1 ,
.Xr wasm-validate 1 ,
.Xr wasm2c 1 ,
.Xr wasm2wat 1 ,
.Xr wast2json 1 ,
.Xr wat-desugar 1 ,
.Xr spectest-interp 1 
.Sh BUGS
If you find a bug, please report it at
.br
.Lk https://github.com/WebAssembly/wabt/issues .


src/tools/wat2wasm.cc
/*
 * Copyright 2016 WebAssembly Community Group participants
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include <cassert>
#include <cstdarg>
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <string>

#include "wabt/config.h"

#include "wabt/binary-writer.h"
#include "wabt/common.h"
#include "wabt/error-formatter.h"
#include "wabt/feature.h"
#include "wabt/filenames.h"
#include "wabt/ir.h"
#include "wabt/option-parser.h"
#include "wabt/resolve-names.h"
#include "wabt/stream.h"
#include "wabt/validator.h"
#include "wabt/wast-parser.h"

using namespace wabt;

static const char* s_infile;
static std::string s_outfile;
static bool s_dump_module;
static int s_verbose;
static WriteBinaryOptions s_write_binary_options;
static bool s_validate = true;
static Features s_features;

static std::unique_ptr<FileStream> s_log_stream;

static const char s_description[] =
    R"(  read a file in the wasm text format, check it for errors, and
  convert it to the wasm binary format.

examples:
  # parse test.wat and write to .wasm binary file with the same name
  $ wat2wasm test.wat

  # parse test.wat and write to binary file test.wasm
  $ wat2wasm test.wat -o test.wasm

  # parse spec-test.wast, and write verbose output to stdout (including
  # the meaning of every byte)
  $ wat2wasm spec-test.wast -v
)";

static void ParseOptions(int argc, char* argv[]) {
  OptionParser parser("wat2wasm", s_description);

  parser.AddOption('v', "verbose", "Use multiple times for more info", []() {
    s_verbose++;
    s_log_stream = FileStream::CreateStderr();
  });
  parser.AddOption('d', "dump-module",
                   "Print a hexdump of the module to stdout",
                   []() { s_dump_module = true; });
  s_features.AddOptions(&parser);
  parser.AddOption('o', "output", "FILE",
                   "Output wasm binary file. Use \"-\" to write to stdout.",
                   [](const char* argument) { s_outfile = argument; });
  parser.AddOption(
      'r', "relocatable",
      "Create a relocatable wasm binary (suitable for linking with e.g. lld)",
      []() { s_write_binary_options.relocatable = true; });
  parser.AddOption(
      "no-canonicalize-leb128s",
      "Write all LEB128 sizes as 5-bytes instead of their minimal size",
      []() { s_write_binary_options.canonicalize_lebs = false; });
  parser.AddOption("debug-names",
                   "Write debug names to the generated binary file",
                   []() { s_write_binary_options.write_debug_names = true; });
  parser.AddOption("no-check", "Don't check for invalid modules",
                   []() { s_validate = false; });
  parser.AddArgument("filename", OptionParser::ArgumentCount::One,
                     [](const char* argument) { s_infile = argument; });

  parser.Parse(argc, argv);
}

static Result WriteBufferToFile(std::string_view filename,
                                const OutputBuffer& buffer) {
  if (s_dump_module) {
    std::unique_ptr<FileStream> stream = FileStream::CreateStdout();
    if (s_verbose) {
      stream->Writef(";; dump\n");
    }
    if (!buffer.data.empty()) {
      stream->WriteMemoryDump(buffer.data);
    }
  }

  if (filename == "-") {
    CHECK_RESULT(buffer.WriteToStdout());
  } else {
    CHECK_RESULT(buffer.WriteToFile(filename));
  }
  return Result::Ok;
}

static std::string DefaultOuputName(std::string_view input_name) {
  // Strip existing extension and add .wasm
  std::string result(StripExtension(GetBasename(input_name)));
  result += kWasmExtension;

  return result;
}

int ProgramMain(int argc, char** argv) {
  InitStdio();

  ParseOptions(argc, argv);

  std::vector<uint8_t> file_data;
  Result result = ReadFile(s_infile, &file_data);
  Errors errors;
  std::unique_ptr<WastLexer> lexer = WastLexer::CreateBufferLexer(
      s_infile, file_data.data(), file_data.size(), &errors);
  if (Failed(result)) {
    WABT_FATAL("unable to read file: %s\n", s_infile);
  }

  std::unique_ptr<Module> module;
  WastParseOptions parse_wast_options(s_features);
  result = ParseWatModule(lexer.get(), &module, &errors, &parse_wast_options);

  if (Succeeded(result) && s_validate) {
    ValidateOptions options(s_features);
    result = ValidateModule(module.get(), &errors, options);
  }

  if (Succeeded(result)) {
    MemoryStream stream(s_log_stream.get());
    s_write_binary_options.features = s_features;
    result = WriteBinaryModule(&stream, module.get(), s_write_binary_options);

    if (Succeeded(result)) {
      if (s_outfile.empty()) {
        s_outfile = DefaultOuputName(s_infile);
      }
      result = WriteBufferToFile(s_outfile.c_str(), stream.output_buffer());
    }
  }

  auto line_finder = lexer->MakeLineFinder();
  FormatErrorsToFile(errors, Location::Type::Text, line_finder.get());

  return result != Result::Ok;
}

int main(int argc, char** argv) {
  WABT_TRY
  return ProgramMain(argc, argv);
  WABT_CATCH_BAD_ALLOC_AND_EXIT
}
